lib/cocoapods-core/cdn_source.rb

```diff
@@ -0,0 +1,210 @@
+        File.open(etag_path, 'w') { |f| f.write(etag_new) } unless etag_new.nil?
+        partial_url
+      end
+    end
+
+    def local_file(partial_url)
+      File.open("#{repo}/#{partial_url}") do |file|
+        yield file if block_given?
+      end
+    end
+
+    def local_file_exist?(partial_url)
+      File.exist?("#{repo}/#{partial_url}")
+    end
+
+    def pod_path_partial(pod_name)
+      pod_path(pod_name).relative_path_from(repo)
+    end
+
+    # @!group Querying the source
+    #-------------------------------------------------------------------------#
+
+    # @return [Array<String>] the list of the name of all the Pods.
+    #
+    #
+    def pods
+      download_file('all_pods.txt')
+      local_file('all_pods.txt', &:to_a)
+    end
+
+    # @return [Array<Version>] all the available versions for the Pod, sorted
+    #         from highest to lowest.
+    #
+    # @param  [String] name
+    #         the name of the Pod.
+    #
+    def versions(name)
+      return nil unless specs_dir
+      raise ArgumentError, 'No name' unless name
+
+      return @versions_by_name[name] unless @versions_by_name[name].nil?
+
+      versions_file_name = pod_path_partial(name).join('index.txt')
+      download_file(versions_file_name)
+
+      loaders = []
+      @versions_by_name[name] ||= local_file(versions_file_name) do |file|
+        file.map do |v|
+          basename = v.chomp
+          # Optimization: ensure all the podspec files at least exist. The correct one will get refreshed
+          # in #specification_path regardless.
+          unless local_file_exist?(pod_path_partial(name).join("#{basename}/#{name}.podspec.json"))
+            loaders << Concurrent::Promise.execute(:executor => @executor) do
+              download_file(pod_path_partial(name).join("#{basename}/#{name}.podspec.json"))
+            end
+          end
+          begin
+            Version.new(basename) if basename[0, 1] != '.'
+          rescue ArgumentError
+            raise Informative, 'An unexpected version directory ' \
+            "`#{basename}` was encountered for the " \
+            "`#{pod_dir}` Pod in the `#{name}` repository."
+          end
+        end
+      end.compact.sort.reverse
+      Concurrent::Promise.zip(*loaders).wait!
+      @versions_by_name[name]
+    end
+
+    # Returns the path of the specification with the given name and version.
+    #
+    # @param  [String] name
+    #         the name of the Pod.
+    #
+    # @param  [Version,String] version
+    #         the version for the specification.
+    #
+    # @return [Pathname] The path of the specification.
+    #
+    def specification_path(name, version)
+      raise ArgumentError, 'No name' unless name
+      raise ArgumentError, 'No version' unless version
+
+      relative_podspec = pod_path_partial(name).join("#{version}/#{name}.podspec.json")
+      download_file(relative_podspec)
+      pod_path(name).join("#{version}/#{name}.podspec.json")
```

Comment:
`"#{version}/#{name}.podspec.json"` could be pulled out into a variable for reuse